### PR TITLE
feat: add yaml feature management

### PR DESCRIPTION
**Pull request by ritave**
This PR builds upon and supersedes #17984. It adds build configuration yaml file (`builds.yml`), and removes hardcoded logic concerning builds - making them all declarative.

Additionally, it adds a feature which I call "Build Features" - it replaces current code fencing based around build types to code fencing around specific features. Each build can contain multiple features. This is necessary because with the growth of MetaMask engineering team there's a lot new pre-release features being build, some are being audited, some aren't, and build types themselves are not enough to maintain that.

Fixes https://app.zenhub.com/workspaces/snaps-615b3a7c08d2b20015eb6c4e/issues/gh/metamask/metamask-planning/502

## Big pictures changes

- [x] Adds `builds.yml` file
- [x] Standarize environmental variable loading order
- [x] Requires all environmental variables to be declared in `builds.yml` before use.
- [x] Allows loading only the environmental variables that are expected to be used during that specific build
- [x] Partial extraction of most existing env variables, some are still modified in-code.
- [x] Build types are now composed of features
- [x] Code fencing is now based on features rather than build types
- [x] Splits flask fencing into desktop and snaps fences
- [x] Removes quite a few build type bugs, especially around `mmi`. Types were hardcoded in multiple places, their behaviour have been varied and was not exactly deterministic. By centralizing all definitions, each build type is now treated the same way.
- [x] Partial support for assets declaration in `builds.yml` rather than hardcoded

## Changes

- `app/*` changes are only code fencing names
- `shared/*` has minor changes to BUILD_TYPES enum and code fencing
- `development/*` has major changes to build system

## Compatibility

The intention of this PR is to maintain backwards compatibility, but due to huge complexity of the change, regressions might have been introduced. Thorough automatic and manual testing is strongly suggested, especially around:

1. **Production builds**
2. Main/Beta/Flask/MMI differences
3. Snaps feature set

### CODE REVIEW

**Review by brad-decker — COMMENTED** [review]
I have been going through file by file and i'm over half way through and haven't found anything worth commenting on. Overall i'm super happy with this change, it makes a lot of sense and makes moving features between build types fairly trivial. 👍🏻 Great work @ritave

File: app/scripts/lib/setupSentry.js [hunk 1]
```diff
@@ -90,7 +89,7 @@ export default function setupSentry({ release, getState }) {
   }
 
   const environment =
-    METAMASK_BUILD_TYPE === BuildType.main
+    METAMASK_BUILD_TYPE === 'main'
```
Comment (brad-decker) [inline app/scripts/lib/setupSentry.js:92]: should we continue to use constants for these strings?

File: app/scripts/metamask-controller.js [hunk 2]
```diff
@@ -840,6 +840,11 @@ export default class MetamaskController extends EventEmitter {
       ],
     });
 
+    const allowLocalSnaps = process.env.ALLOW_LOCAL_SNAPS;
+    const requireAllowlist = process.env.REQUIRE_SNAPS_ALLOWLIST;
+    assert(typeof allowLocalSnaps === 'boolean');
+    assert(typeof requireAllowlist === 'boolean');
```
Comment (brad-decker) [inline app/scripts/metamask-controller.js]: Could we possibly move these assertions to higher up in the execution layer? It feels weird to be doing assertions on build variables in the application logic.

**Review by brad-decker — CHANGES_REQUESTED** [review]

File: app/scripts/metamask-controller.js (hunk 2 above)
Comment (FrederikBolding) [reply app/scripts/metamask-controller.js]: Yeah, I removed these assertions for now. They also broke a bunch of tests

File: app/scripts/lib/setupSentry.js (hunk 1 above)
Comment (FrederikBolding) [reply app/scripts/lib/setupSentry.js:92]: We could keep the enum, but there is no guarantee it matches what is defined in `builds.yml`. Would you prefer that?

**Comment by FrederikBolding** [comment]
@brad-decker This now passes CI and is ready for more review 👍

File: app/scripts/lib/setupSentry.js (hunk 1 above)
Comment (FrederikBolding) [reply app/scripts/lib/setupSentry.js:92]: I'll leave that up to you, but in general we don't seem to check it that much after this refactor. Which seems like a more sensible approach anyway

File: development/build/transforms/README.md [hunk 3]
```diff
@@ -49,7 +49,7 @@ Note that multiple build types can be specified by separating them with
 commands inside the parameter parentheses:
 
 ```javascript
-///: BEGIN:ONLY_INCLUDE_IN(beta,flask)
+///: BEGIN:ONLY_INCLUDE_IN(build-beta,flask)
```
Comment (cryptotavares) [inline development/build/transforms/README.md:52]: Should be `build-flask` as well

File: app/scripts/lib/setupSentry.js (hunk 1 above)
Comment (ritave) [reply app/scripts/lib/setupSentry.js:92]:
```suggestion
    METAMASK_BUILD_TYPE === process.env.METAMASK_BUILD_TYPE_DEFAULT
```

File: development/build/transforms/README.md (hunk 3 above)
Comment (FrederikBolding) [reply development/build/transforms/README.md]: Fixed!

File: test/jest/setup.js [hunk 4]
```diff
@@ -3,6 +3,8 @@
 import nock from 'nock';
 import '@testing-library/jest-dom';
 
+process.env.SUPPORT_LINK = 'https://support.metamask.io';
```
Comment (cryptotavares) [inline test/jest/setup.js:6]: Can we move this to `test/jest/env.js`... I belive that is where we want to setup all jest tests specific env vars.

**Comment by cryptotavares** [comment]
Looks great! I really love how easy it will be to move features across build types 😎 

The only reservation that I have is naming it features [here](https://github.com/MetaMask/metamask-extension/pull/18125/files#diff-d6a9a173a5b80ebccbd44d172db70d11715731b4ebb89d44465d3289da1bd642R90).
Code fencing, though useful, introduces a lot of complexity. Developers are required to take into account code blocks that may or may not exist for a specific build type.
By:
1. removing a lot of friction to add new code fences
2. detach code fences from build types
3. couple it to features (and name it features)

I can see those proliferating and making the code base becoming exponentially more complex and hard to change.

I believe that this could be partially mitigated if we re-named `features` from the build.yml into `fenced-features`? (I always prefer being explicit). Finally we should also add a markdown to the docs about adding `fenced-features` and when that should be considered (to help out on the decision process and avoid code fencing proliferation).
IMO those small changes could make huge difference.

What do you think?

**Comment by brad-decker** [comment]
@FrederikBolding @cryptotavares i'm not in favor of renaming the key at this point, though I acknowledge that we don't want to increase the amount of code fencing happening. Perhaps this issue can be handled through education and documentation rather than making the keys slightly more obscurely named?

**Comment by FrederikBolding** [comment]
@brad-decker Rebased!

**Comment by FrederikBolding** [comment]
Sorry - had a few issues after rebasing.

**Review by pedronfigueiredo — DISMISSED** [review]

**Comment by pedronfigueiredo** [comment]
LGTM

**Review by brad-decker — DISMISSED** [review]
LGTM

**Review by brad-decker — APPROVED** [review]

**Review by pedronfigueiredo — APPROVED** [review]

**Pull request merged**